js/data/bucket/symbol_bucket.js

```diff
@@ -184,6 +192,12 @@ class SymbolBucket {
             // It's better to place labels on one long line than on many short segments.
             this.features = mergeLines(this.features);
         }
+
+        // set up this mapping to allow for constant-time lookups in place()
+        this.featuresByIndex = {};
```

Comment:
Sure; if we're doing that I'd say we should go with 'replace' over 'adjust': adjusting it to these requirements would mean changing it from a StructArray, thus making it non-transferrable and no longer suitable for the cross-source placement project...  if that's going to be the case, I don't see much reason to hold on to this state (`symbolInstancesArray`, `symbolQuadsArray`) at the WorkerTile level.

I'm good to go ahead with this, but I just want to flag that dropping symbolInstancesArray will make the changes in this PR a notch more invasive, removing some stuff that we may end up having to partially redo later when we tackle cross-source placement -- but it's probably true that even if we left it in place, making it work with data-driven styling would anyway have caused significant changes to what's there now.